### PR TITLE
[CH] Shuffle writer connects to CH pipeline

### PR DESCRIPTION
**Pull request by liuneng1994**
## What changes were proposed in this pull request?

shuffle writer现在可以作为Processor接入ClickHouse pipeline当中。
image
在fallback模式下，会在jni中以循环的形式完成stage的执行，主要原因是，某些情况下的fallback会有spark的whole code gen，
其中code gen生成的代码会使用TaskContext，需要保证执行线程为task线程
image

移除了CachedShuffleWriter，新的SparkExchangeSink行为与原有shuffleWriter保持一致

同时还做了一下改动：
* 支持native的inputFileName,InputBlockStart, InputBlockLength
* shuffle Wall time统计，在Processor层面统计完整的shuffle耗时
* LocalExecutor移出SerilizedPlanParser
* DefaultHashAggregateResultStep与DefaultHashAggregateResultTransform的output header不匹配问题

## How was this patch tested?

unit tests

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by liuneng1994** [comment]
Run Gluten Clickhouse CI

**Comment by liuneng1994** [comment]
Run Gluten Clickhouse CI

**Comment by liuneng1994** [comment]
image
性能测试正常，无回退

File: gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala [hunk 1]
```diff
@@ -332,6 +332,8 @@ case class OffloadProject() extends OffloadSingleNode with LogLevelUtil {
           expr => rewriteExpr(expr, replacedExprs).asInstanceOf[NamedExpression]
         }
         val newChild = addMetadataCol(projectExec.child, replacedExprs)
+        // tag child again, addMetadataCol lost the tag.
+        addHint.apply(newChild)
```
Comment (liuneng1994) [inline gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala:336]: @zhztheplayer please help review this code, fixed an issue where fallback could not be performed correctly

**Comment by liuneng1994** [comment]
@kerwin-zk I modified some of the CelebornShuffleManager code. Please review it when you have time. Thanks

**Comment by liuneng1994** [comment]
relate with https://github.com/apache/incubator-gluten/pull/6517

**Review by kerwin-zk — COMMENTED** [review]
@liuneng1994 The changes related to Celeborn are fine.

File: cpp-ch/local-engine/Storages/SourceFromJavaIter.cpp [hunk 2]
```diff
@@ -102,10 +105,11 @@ DB::Chunk SourceFromJavaIter::generate()
     SCOPE_EXIT({CLEAN_JNIENV});
 
     DB::Block * input_block = nullptr;
-    if (first_block) [[unlikely]]
+    bool clean_first_block = false;
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/SourceFromJavaIter.cpp:108]: Is this necessary ? just let `first_block = std::nullopt` at last in anyway

File: cpp-ch/local-engine/Parser/InputFileNameParser.cpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        const auto & column = header.getByPosition(i);
+        if (column.name == INPUT_FILE_NAME)
+        {
+            if (!file_name.has_value())
+                throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Input file name is not set");
+            auto type_string = std::make_shared<DB::DataTypeString>();
+            auto file_name_column = type_string->createColumn();
+            file_name_column->insertMany(file_name.value(), chunk.getNumRows());
+            output_columns.insert(output_columns.begin() + i, std::move(file_name_column));
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/InputFileNameParser.cpp:168]: use `output_columns.push_back` ?

File: cpp-ch/local-engine/Parser/InputFileNameParser.cpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+bool InputFileNameParser::containsInputFileColumns(const DB::Block & block)
+{
+    return hasInputFileNameColumn(block) || hasInputFileBlockStartColumn(block) || hasInputFileBlockLengthColumn(block);
+}
+
+DB::Block InputFileNameParser::removeInputFileColumn(const DB::Block & block)
+{
+    const auto & columns = block.getColumnsWithTypeAndName();
+    DB::ColumnsWithTypeAndName result_columns;
+    for (const auto & column : columns) {
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/InputFileNameParser.cpp:201]: format `{`

File: cpp-ch/local-engine/Parser/InputFileNameParser.cpp [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    }
+    return result_columns;
+}
+
+void InputFileNameParser::addInputFileProjectStep(DB::QueryPlan & plan)
+{
+    if (!file_name.has_value() && !block_start.has_value() && !block_length.has_value()) return;
+    auto step = std::make_unique<InputFileExprProjectStep>(plan.getCurrentDataStream(), file_name, block_start, block_length);
+    step->setStepDescription("Input file expression project");
+    plan.addStep(std::move(step));
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/InputFileNameParser.cpp:215]: Are this step's metrics reported to gluten ?

File: cpp-ch/local-engine/Parser/InputFileNameParser.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    for (size_t i = 0; i < header.columns(); ++i)
+    {
+        const auto & column = header.getByPosition(i);
+        if (column.name == INPUT_FILE_NAME)
+        {
+            if (!file_name.has_value())
+                throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Input file name is not set");
+            auto type_string = std::make_shared<DB::DataTypeString>();
+            auto file_name_column = type_string->createColumn();
+            file_name_column->insertMany(file_name.value(), chunk.getNumRows());
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/InputFileNameParser.cpp:167]: This should be a const column. Insert mutli same values into a new empty column is not efficient

File: cpp-ch/local-engine/Parser/InputFileNameParser.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+namespace DB
+{
+    class Chunk;
+}
+
+namespace local_engine
+{
+    class InputFileNameParser
+    {
+    public:
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/InputFileNameParser.h:29]: code format style

File: cpp-ch/local-engine/Shuffle/SparkExchangeSink.h [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    SplitResult getSplitResultCopy() const
+    {
+        return split_result;
+    }
+
+    DB::Block getOutputHeaderCopy() const
+    {
+        return output_header.cloneEmpty();
+    }
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Shuffle/SparkExchangeSink.h:65]: return const references, if we don't need to worry the content be modified in `SparkExechangeSink` during usage outside

File: cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    if (rss_pusher)
+    {
+        GET_JNIENV(env)
+        jclass celeborn_partition_pusher_class =
+            CreateGlobalClassReference(env, "Lorg/apache/spark/shuffle/CelebornPartitionPusher;");
+        jmethodID celeborn_push_partition_data_method =
+            GetMethodID(env, celeborn_partition_pusher_class, "pushPartitionData", "(I[BI)I");
+        CLEAN_JNIENV
+        celeborn_client = std::make_unique<CelebornClient>(rss_pusher, celeborn_push_partition_data_method);
+        use_rss = true;
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp:116]: could remove `use_rss`, just use `celeborn_client != nullptr`

File: cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+    split_result.partition_lengths.resize(options.partition_num, 0);
+    split_result.raw_partition_lengths.resize(options.partition_num, 0);
+}
+
+std::shared_ptr<PartitionWriter> createPartitionWriter(const SplitOptions& options, bool use_sort_shuffle, std::unique_ptr<CelebornClient> celeborn_client)
+{
+    if (celeborn_client)
+    {
+        if (use_sort_shuffle)
+            return std::make_shared<MemorySortCelebornPartitionWriter>(options, std::move(celeborn_client));
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp:138]: I wonder whether a celeborn client could be shared among multi threads. If it's sharable, make it a shared_ptr may be better.

File: cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    {
+        partition_writers[i] = createPartitionWriter(options, use_sort_shuffle, std::move(celeborn_client));
+        sinks[i] = std::make_shared<SparkExchangeSink>(input_header, partitioner_creator(options), partition_writers[i], output_columns_indicies, use_sort_shuffle);
+    }
+}
+
+void SparkExchangeManager::setSinksToPipeline(DB::QueryPipelineBuilder & pipeline) const
+{
+    size_t count = 0;
+    Pipe::ProcessorGetterWithStreamKind getter = [&](const Block & header, Pipe::StreamType stream_type) -> ProcessorPtr
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp:164]: lost namespace `DB`?

File: cpp-ch/local-engine/Parser/LocalExecutor.h [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+struct SparkBuffer
+{
+    char * address;
+    size_t size;
+};
+
+class LocalExecutor : public BlockIterator
+{
+public:
+    static LocalExecutor * getCurrentExecutor() { return current_executor; }
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/LocalExecutor.h:38]: Worry about making this a global variable? Could explain it lifetime?

File: cpp-ch/local-engine/Parser/InputFileNameParser.cpp (hunk 3 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Parser/InputFileNameParser.cpp:168]: Here need to insert the newly added columns into the corresponding positions, which can only be done with insert

File: cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp (hunk 9 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp:116]: celeborn cannot bring performance improvement in parallel writing. In the case of rss, celeborn client will be moved to the Sink object, so an additional variable is needed

File: cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp (hunk 10 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Shuffle/SparkExchangeSink.cpp:138]: celeborn will not have any performance benefits in parallel, and will later resize the pipeline to ensure that there is only one sink

File: cpp-ch/local-engine/Parser/LocalExecutor.h (hunk 12 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Parser/LocalExecutor.h:38]: add comment

File: cpp-ch/local-engine/local_engine_jni.cpp [hunk 13]
```diff
@@ -535,10 +537,55 @@ JNIEXPORT void Java_org_apache_gluten_vectorized_CHStreamReader_nativeClose(JNIE
...
+        jobject iter,
+        const String & name,
+        const local_engine::SplitOptions& options,
+        jobject rss_pusher = nullptr
+        )
+{
+    auto * current_executor = local_engine::LocalExecutor::getCurrentExecutor();
+    chassert(current_executor);
+    local_engine::SplitterHolder * splitter = nullptr;
+    // handle fallback, whole stage fallback or partial fallback
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/local_engine_jni.cpp:550]: In which cases， current_executor will be nullptr or not? It's better to add a comment here

Comment (liuneng1994) [reply cpp-ch/local-engine/local_engine_jni.cpp]: done

**Comment by lgbo-ustc** [comment]
LGTM

**Review by lgbo-ustc — APPROVED** [review]

**Pull request merged**